### ISSUE TITLE
Sane default for editor

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#41]
We should have some sane default (rather than `vim`) for `general -> editor`.

Some ideas:
- Use `QDesktopServices` to decide.
- Maybe there's some XDG tool.
- notepad.exe on Windows
- Some other default editor on Mac
- List of editors to try?
- `$EDITOR` and `$VISUAL` probably won't work as they aren't GUI editors typically

**Comment by ersi-dnd** [comment]
Can you tell what kind of external editor settings work for you? For me the default doesn't work and nothing else works either, no matter how I change it. I'm trying to get nano open up in xterm.

**Comment by Carpetsmoker** [comment]
@ersi-dnd This should work:

    xterm -e nano "{}"

**Comment by ersi-dnd** [comment]
Yep, that's what I thought too. I always get the error 

open-editor: Unrecognised arguments "{}"

**Comment by Carpetsmoker** [comment]
Are you sure you used the *exact* command as above? In particular, the single quotes are important

**Comment by ersi-dnd** [comment]
Yes, after about a thousandth try, I am absolutely sure. I've been doing this for at least a month really. I have tried editing the .conf file directly, the :settings page, and :set from the address bar, with and without the curly braces, and more. And the default is not working either, gives the same error. So, probably something to do with curly braces or quotes in my system.

By the way, yesterday I discovered that system locale has an effect on how regex works. It's to do with different order of letters in different alphabets. Maybe system locale has an effect on how curly braces are treated too? :)

**Comment by The-Compiler** [comment]
@ersi-dnd Can we please move this kind of thing to either the [mailinglist](https://lists.schokokeks.org/mailman/listinfo.cgi/qutebrowser) or the IRC channel (`#qutebrowser` on Freenode) - or at the very least, can you open a new issue? This issue literally has *nothing* to do with your problem.

**Comment by lamarpavel** [comment]
Since this browser is closely related to vim, wouldn't it be worth a try to actually make use of vim within the browser? Instead of spawning an external editor, start vim as a server and relay the input to qute to vim and vims output back to qute. qutebrowser would then essentially be an interface to vim whenever you enter text on a website.

There is a plugin for Firefox doing that, but I can't remember the name right now. I do remember that it was rather clumsy and not performing well, but that might have been due to integration as plugin in Firefox's weird plugin architecture and might work really well when natively implemented in qute.

Alternatively, qute could use neovim for that, which supplies an RPC interface for exactly that. The problem here being, that neovim has not yet reached a stable release and the RPC-interface might still change.

I would find that solution to be much more comfortable than spawning an external editor.

**Comment by ersi-dnd** [comment]
Negative. I happen to like vim-like browsers, but not at all vi(m) itself. The only thing I want to know about it is how to get out of it.

**Comment by The-Compiler** [comment]
@lamarpavel This is certainly an interesting idea - or if that doesn't work out, some kind of vim-emulation for form fields might be useful as well. Can you open a separate issue for that too, please? After all, this issue only is about auto-detecting what editor should be used :wink:

**Comment by Ambrevar** [comment]
Some programs such as git and mutt (if I'm not mistaken) proceed as follows:
- Look at `$VISUAL` first.
- If nothing was found, look at `$EDITOR`.

Others only lookup `$EDITOR`.

That last point allows me to separate the use cases:
- Set $VISUAL to a graphical editor: useful when the caller does not handle the spawning of a new window.
- Set $EDITOR to a console editor: useful when running a terminal and you don't want to spawn a new window.

I think Qutebrowser could definitely make use of those variables as they are widespread. There is nothing _by design_ that prevents them from being used outside a terminal.
